framework/DBDD_optimized.sage

```diff
@@ -0,0 +1,210 @@
+from fpylll import *
+from fpylll.algorithms.bkz2 import BKZReduction
+
+load("../framework/load_strategies.sage")
+load("../framework/DBDD.sage")
+load("../framework/proba_utils.sage")
+
+def need_lattices_bases(fn):
+    """
+    Decorator which removes linear dependencies in lattice generator sets
+    of the DBDD instance to manipulate bases in the execution of the
+    decorated method
+    """
+    def decorated(self, *args, **kwargs):
+        has_been_reduced = False
+
+        # Remove linear dependencies in the primal lattice generator set
+        if self.B is not None:
+            has_been_reduced = has_been_reduced or (self.B.nrows() > self._dim)
+            self.B = eliminate_linear_dependencies(self.B, dim=self._dim)
+
+        # Remove linear dependencies in the dual lattice generator set
+        if self.D is not None:
+            if has_been_reduced and (self.D.nrows() > self._dim):
+                # Display a warning because it does the calculus twice
+                # Maybe there exists a method to use the previous calculation
+                #   to remove dependencies in the dual basis ?
+                self.logging("Double computation with LLL", priority=0, style='WARNING', newline=False)
+            has_been_reduced = has_been_reduced or (self.D.nrows() > self._dim)
+            self.D = eliminate_linear_dependencies(self.D, dim=self._dim)
+
+        # And then, execute the function with bases (and not just generator sets)
+        return fn(self, *args, **kwargs)
+    return decorated
+
+
+class DBDD_optimized(DBDD):
+    """
+    This class defines all the elements defining a DBDD instance with all
+    the basis computations, with some performance optimizations
+    """
+
+    def __init__(self, B, S, mu, u=None, verbosity=1, homogeneous=False, float_type="ld", D=None, Bvol=None):
+        assert B or D
+        self._dim = (B or D).nrows() # Lattice dimension
+        self.Pi = identity_matrix(self._dim) # Reduction matrix
+        self.Gamma = identity_matrix(self._dim) # Substitution matrix
+        super().__init__(B, S, mu, u=u, verbosity=verbosity, homogeneous=homogeneous, float_type=float_type, D=D, Bvol=None)
```

Comment:
Can you expend a bit on the role of Pi and Gamma ?

Reply:
I later see that a *solution* can be "reduced", and I am afraid that you might be using "reduction" to mean "projection". Can you clarify ?

Reply:
My notion of reduction is not exactly the notion of projection, but not far. Indeed, by definition, a projection is a linear transformation from a vector space to itself. But here, the output vector space is not the same vector space, but it is a vector space with a smaller dimension. And it is the idea of this optimization, we want to decrease the dimension of the lattice vectors to accelerate the computations. I called the _reduced lattice_ the lattice where I decrease the vector space dimension.

Reply:
Do you want I rename the process, or I just expend the role of Pi and Gamma by explaining that I said just before?

Reply:
I think "reduced" is too confusing: LLL does lattice reduction, we can't reuse that term. I think projection is ok even if not an endomorphism. I see people in algebra project groups on other groups all the time. I'd really really prefer projection. So I would suggest to do both: rename and expend :)

Reply:
I understand the problem with the term "reduction". But I don't like the term "projection", because the idea of my "reduction" is just to reorganize the content of the lattice into a smaller lattice where the notion of "projection" often involves a loss of information. So, I pushed a commit with the new term of "Lattice Embedding".